src/Monolog/Formatter/GoogleCloudLoggingFormatter.php

```diff
@@ -32,9 +36,49 @@ protected function normalizeRecord(LogRecord $record): array
         $normalized['severity'] = $normalized['level_name'];
         $normalized['time'] = $record->datetime->format(DateTimeInterface::RFC3339_EXTENDED);
 
+        // Tag with Trace ID for request attribution
+        $normalized['logging.googleapis.com/trace'] = $this->getTraceID();
+
         // Remove keys that are not used by GCP
         unset($normalized['level'], $normalized['level_name'], $normalized['datetime']);
 
         return $normalized;
     }
+
+    private function getTraceID(): ?string
+    {
+        if (empty($this->traceID) && !empty($_SERVER['HTTP_X_CLOUD_TRACE_CONTEXT'])) {
+            $matched = preg_match(
+                self::CONTEXT_HEADER_FORMAT,
+                $_SERVER['HTTP_X_CLOUD_TRACE_CONTEXT'] ?? '',
+                $matches,
+            );
+
+            if (!$matched) {
+                return null;
+            }
+
+            $projectID = $this->getProjectID();
+            if (empty($projectID)) {
+                return null;
+            }
+
+            $this->traceID = 'projects/'.$projectID.'/traces/'.strtolower($matches[1]);
+        }
+
+        return $this->traceID;
+    }
+
+    private function getProjectID(): ?string
+    {
+        if (isset($_SERVER['GOOGLE_CLOUD_PROJECT'])) {
+            return $_SERVER['GOOGLE_CLOUD_PROJECT'];
+        }
+
+        if (class_exists('\Google\Cloud\Core\Compute\Metadata')) {
+            return (new \Google\Cloud\Core\Compute\Metadata())->getProjectId();
```

Comment:
also, it looks like the implementation of `Google\Cloud\Core\Compute\Metadata::getProjectId()` [does not do caching](https://github.com/googleapis/google-cloud-php-core/blob/686ffd0bb1328cd2ee963c4444e2ab4b111ba3ba/src/Compute/Metadata/Readers/HttpHandlerReader.php#L51), so maybe it's better to retrieve the project id just once at the constructor of this `GoogleCloudLoggingFormatter` as not to incur HTTP round trip every time there is a new log entry to format.

Reply:
You'll hopefully see further up, we set the static property `$this->traceID` and once set, we always return that static value.

The project ID is only retrieved to construct that value the single time it is constructed, so hopefully this is already accounted for.

Reply:
Caching `$this->traceID` may be fine if it's guaranteed that `GoogleCloudLoggingFormatter` is always created new for each request. I think this is the case with PHP app with nginx + php-fpm but may not be the case with other kind of deployment (like with Laravel Octane where objects are created just once when the app starts and reused across multiple requests).

I am not familiar enough with monolog + octane to confidently say whether Octane reuses the formatter object across different requests, but to be safe I would not cache the trace id but only the project id.

Reply:
Yeah it's probably safer not to cache, as retrieving the trace id doesn't seem that expensive.

Reply:
And I would definitely not cache this as a static property. Cache on an instance property then a new instance would at least refetch from a clean state. Static property will definitely cause issues in some envs.